byzcoin/secure_darc_test.go

```diff
@@ -0,0 +1,195 @@
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestSecureDarc(t *testing.T) {
+	local := onet.NewTCPTest(cothority.Suite)
+	defer local.CloseAll()
+
+	signer := darc.NewSignerEd25519(nil, nil)
+	_, roster, _ := local.GenTree(3, true)
+
+	genesisMsg, err := DefaultGenesisMsg(CurrentVersion, roster, []string{}, signer.Identity())
+	require.Nil(t, err)
+	gDarc := &genesisMsg.GenesisDarc
+	genesisMsg.BlockInterval = time.Second
+	cl, _, err := NewLedger(genesisMsg, false)
+	require.Nil(t, err)
+
+	restrictedSigner := darc.NewSignerEd25519(nil, nil)
+	unrestrictedSigner := darc.NewSignerEd25519(nil, nil)
+
+	log.Info("spawn a new secure darc with spawn:insecure_darc - fail")
+	secDarc := gDarc.Copy()
+	require.NoError(t, secDarc.Rules.AddRule("spawn:insecure_darc", []byte(restrictedSigner.Identity().String())))
+	secDarcBuf, err := secDarc.ToProto()
+	require.NoError(t, err)
+	ctx := ClientTransaction{
+		Instructions: []Instruction{{
+			InstanceID: NewInstanceID(gDarc.GetBaseID()),
+			Spawn: &Spawn{
+				ContractID: ContractSecureDarcID,
+				Args: []Argument{{
+					Name:  "darc",
+					Value: secDarcBuf,
+				}},
+			},
+			SignerCounter: []uint64{1},
+		}},
+	}
+	require.Nil(t, ctx.SignWith(signer))
+	_, err = cl.AddTransactionAndWait(ctx, 10)
+	require.Error(t, err)
+
+	log.Info("do the same but without spawn:insecure_darc - pass")
+	require.NoError(t, secDarc.Rules.DeleteRules("spawn:insecure_darc"))
+	require.NoError(t, secDarc.Rules.UpdateRule("invoke:secure_darc."+cmdDarcEvolve, []byte(restrictedSigner.Identity().String())))
+	require.NoError(t, secDarc.Rules.UpdateRule("invoke:secure_darc."+cmdDarcEvolveUnrestriction, []byte(unrestrictedSigner.Identity().String())))
+	secDarcBuf, err = secDarc.ToProto()
+	require.NoError(t, err)
+	ctx = ClientTransaction{
+		Instructions: []Instruction{{
+			InstanceID: NewInstanceID(gDarc.GetBaseID()),
+			Spawn: &Spawn{
+				ContractID: ContractSecureDarcID,
+				Args: []Argument{{
+					Name:  "darc",
+					Value: secDarcBuf,
+				}},
+			},
+			SignerCounter: []uint64{1},
+		}},
+	}
+	require.Nil(t, ctx.SignWith(signer))
+	_, err = cl.AddTransactionAndWait(ctx, 10)
+	require.NoError(t, err)
+
+	log.Info("evolve to add rules - fail")
+	{
+		secDarc2 := secDarc.Copy()
+		require.NoError(t, secDarc2.EvolveFrom(secDarc))
+		secDarc2.Rules.AddRule("spawn:coin", secDarc.Rules.Get("invoke:secure_darc."+cmdDarcEvolveUnrestriction))
+		secDarc2Buf, err := secDarc2.ToProto()
+		ctx2 := ClientTransaction{
+			Instructions: []Instruction{{
+				InstanceID: NewInstanceID(secDarc.GetBaseID()),
+				Invoke: &Invoke{
+					ContractID: ContractSecureDarcID,
+					Command:    cmdDarcEvolve,
+					Args: []Argument{{
+						Name:  "darc",
+						Value: secDarc2Buf,
+					}},
+				},
+				SignerCounter: []uint64{1},
+			}},
+		}
+		require.Nil(t, ctx2.SignWith(restrictedSigner))
+		_, err = cl.AddTransactionAndWait(ctx2, 10)
+		require.Error(t, err)
+	}
+
+	log.Info("evolve to modify existing rules - pass")
```

Comment:
Is there a test for "I am restricted, but I want to remove the restriction on myself -> fail"?
Well, in writing it down, it seems covered by other tests and probably not required, but please see if my question has shaken lose any ideas for you...

Reply:
it's a good idea, I added a test for it and the test caught a bug